### PR TITLE
feat(react-query): add support for async iterables in mutations

### LINKED ISSUE #5846: feat: support for Streaming responses to be reflected in useMutation data
### Describe the feature you'd like to request

Hi, I'm very grateful that trpc supports streaming!

I have read the HTTP Batch Stream Link documentation and implemented the query as it was written and was able to confirm that useQuery works fine.

However, when I implement mutation as well, the value of data in useMutation is AsyncGenerator.

```ts
// @filename: server.ts
import { publicProcedure, router } from './trpc';
 
const appRouter = router({
  examples: {
    iterable: publicProcedure.query(async function* () {
      for (let i = 0; i < 3; i++) {
        await new Promise((resolve) => setTimeout(resolve, 500));
        yield i;
      }
    }),
    iterable2: publicProcedure.mutation(async function* () {
      for (let i = 0; i < 10; i++) {
        await new Promise((resolve) => setTimeout(resolve, 500));
        yield i;
      }
    }),
  },
});
 
export type AppRouter = typeof appRouter;
 
 
// @filename: client.tsx
import { createTRPCClient, unstable_httpBatchStreamLink } from '@trpc/client';
import type { AppRouter } from './server';
 
const trpc = createTRPCClient<AppRouter>({
  links: [
    unstable_httpBatchStreamLink({
      url: 'http://localhost:3000',
    }),
  ],
});

// iterable.data is number[] | undefined
const iterable = trpc.examples.iterable.useQuery();
 
// iterable2.mutate is AsyncGenerator<number, void, unknown> | undefined
const iterable2 = trpc.examples.iterable2.useMutation();
```

I would like useMutation to reflect the streaming response in the data as well as useQuery.

### Describe the solution you'd like to see

I checked the https://github.com/trpc/trpc/pull/5741 to see if I could somehow determine the cause myself, but I could not figure out what the difference is between query and mutation.

I tried to change useMutation as well as useQuery but could not get it to work.
Please see comments on code.

```ts
    const hook = __useMutation(
      {
        ...opts,
        mutationKey: mutationKey,
        mutationFn: async (input) => {
          const result = await client.mutation(
            ...getClientArgs([path, { input }], opts),
          );

          if (isAsyncIterable(result)) {

            // is this correct?
            const mutationCache = queryClient.getMutationCache();
            const mutation = mutationCache.build(queryClient, {
              mutationKey,
            });

            // error: setState is not exist
            await mutation.setState({
              data: [],
              status: 'success',
            });

            const aggregate: unknown[] = [];
            for await (const value of result) {
              aggregate.push(value);

              // error: setState is not exist
              mutation.setState({
                data: [...aggregate],
              });
            }
            return aggregate;
          }
          return result;
        },
        onSuccess(...args) {
          const originalFn = () =>
            opts?.onSuccess?.(...args) ?? defaultOpts?.onSuccess?.(...args);

          return mutationSuccessOverride({
            originalFn,
            queryClient,
            meta: opts?.meta ?? defaultOpts?.meta ?? {},
          });
        },
      },
      queryClient,
    ) as UseTRPCMutationResult<unknown, TError, unknown, unknown>;
```

### Describe alternate solutions

As a workaround, it is still possible to handle AsyncGenerator on the client side as follows.

```.tsx
  const [text, setText] = useState<string>('');
  const iterable = trpc.examples.iterable.useMutation();
  const callMutation = async () => {
    iterable.mutate(
      { text: inputText },
      {
        onSuccess: async (data) => {
          for await (const val of data) {
            setText((prev) => prev + val);
          }
        },
      },
    );
  };
```

### Additional information

Discord Link: https://discord.com/channels/867764511159091230/1156727678267822120/1256862319691042846

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

### PR DESCRIPTION
**Pull request by KATT**
Closes #5846

## 🎯 Changes

Not done, need feedback

## ✅ Checklist

- [ ] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [ ] If necessary, I have added documentation related to the changes made.
- [ ] I have added or updated the tests related to the changes made.

### CODE REVIEW

File: packages/react-query/src/shared/hooks/createHooksInternal.tsx [hunk 1]
```diff
@@ -273,12 +273,37 @@ export function createRootHooks<
...
+            for await (const value of result as AsyncIterable<unknown>) {
+              setDataOverride((prev) => {
+                const next = [...(prev ?? []), value];
+                return next;
+              });
+              yield value;
+            }
+          }
+
+          return iterable();
```
Comment (KATT) [inline packages/react-query/src/shared/hooks/createHooksInternal.tsx:306]: cc @TkDodo  👀 

is there a way for me to make the mutation stay as _pending_ but still have `mutateAsync` return the result?

tl;dr

```ts
const mutation = trpc.iterable.useMutation()

mutation.data // <--- should be an array that gets updated

// [...]
await mutation.mutateAsync(); // <--- should return an `AsyncIterable`
```

I'm not sure how to access the `MutationObserver` within the function here

Comment (TkDodo) [reply packages/react-query/src/shared/hooks/createHooksInternal.tsx:306]: I don't think so. pending state is defined as "we have no data".

> I'm not sure how to access the MutationObserver within the function here

It's not exposed. You would probably have to roll your own `useMutation`. If you look at the code, it doesn't do that much: Basically create an observer, subscribe to it in `useSyncExternalStore` and return the result ...
